cmd/containers-storage/create.go

```diff
@@ -22,14 +25,124 @@ var (
 	paramMetadata     = ""
 	paramMetadataFile = ""
 	paramCreateRO     = false
+	paramHostUIDMap   = false
+	paramHostGIDMap   = false
+	paramUIDMap       = ""
+	paramGIDMap       = ""
+	paramSubUIDMap    = ""
+	paramSubGIDMap    = ""
 )
 
+func paramIDMapping() (*storage.IDMappingOptions, error) {
+	options := storage.IDMappingOptions{
+		HostUIDMapping: paramHostUIDMap,
+		HostGIDMapping: paramHostGIDMap,
+	}
+	if paramHostUIDMap && (len(paramUIDMap) > 0 || paramSubUIDMap != "") {
+		return nil, fmt.Errorf("host uid map specified along with UID mapping")
+	}
+	if paramHostGIDMap && (len(paramGIDMap) > 0 || paramSubGIDMap != "") {
+		return nil, fmt.Errorf("host gid map specified along with GID mapping")
+	}
+	if len(paramUIDMap) > 0 && paramSubUIDMap != "" {
+		return nil, fmt.Errorf("subuid map specified along with UID mapping")
+	}
+	if len(paramGIDMap) > 0 && paramSubGIDMap != "" {
+		return nil, fmt.Errorf("subgid map specified along with GID mapping")
+	}
+	if paramSubUIDMap != "" && paramSubGIDMap == "" {
+		return nil, fmt.Errorf("subuid map specified but no subgid map")
+	}
+	if paramSubUIDMap == "" && paramSubGIDMap != "" {
+		return nil, fmt.Errorf("subgid map specified but no subuid map")
+	}
+	nonDigitsToWhitespace := func(r rune) rune {
+		if strings.IndexRune("0123456789", r) == -1 {
+			return ' '
+		} else {
+			return r
+		}
+	}
+	parseTriple := func(spec []string) (container, host, size uint32, err error) {
+		cid, err := strconv.ParseUint(spec[0], 10, 32)
+		if err != nil {
+			return 0, 0, 0, fmt.Errorf("error parsing id map value %q: %v", spec[0], err)
+		}
+		hid, err := strconv.ParseUint(spec[1], 10, 32)
+		if err != nil {
+			return 0, 0, 0, fmt.Errorf("error parsing id map value %q: %v", spec[1], err)
+		}
+		sz, err := strconv.ParseUint(spec[2], 10, 32)
+		if err != nil {
+			return 0, 0, 0, fmt.Errorf("error parsing id map value %q: %v", spec[2], err)
+		}
+		return uint32(cid), uint32(hid), uint32(sz), nil
+	}
+	if len(paramUIDMap) > 0 {
+		uidSpec := strings.Fields(strings.Map(nonDigitsToWhitespace, paramUIDMap))
+		if len(uidSpec)%3 != 0 {
+			return nil, fmt.Errorf("uid map is malformed")
+		}
+		for i := range uidSpec {
+			if i%3 != 0 {
+				continue
+			}
+			cid, hid, size, err := parseTriple(uidSpec[i : i+3])
+			if err != nil {
+				return nil, fmt.Errorf("uid map is malformed")
+			}
+			mapping := idtools.IDMap{
+				ContainerID: int(cid),
+				HostID:      int(hid),
+				Size:        int(size),
+			}
+			options.UIDMap = append(options.UIDMap, mapping)
+		}
+	}
+	if len(paramGIDMap) > 0 {
+		gidSpec := strings.Fields(strings.Map(nonDigitsToWhitespace, paramGIDMap))
```

Comment:
(The triple parsing code might be big enough to extract to a function shared for `paramUIDMap` and `paramGIDMap`.  The way `nonDigitsToWhitespace` throws away information worries me a bit.)

Reply:
Factoring that out helped with readability.  I'm cheating with `nonDigestToWhitespace()` so that we don't have to enforce use of particular separators for the list of mappings if there's more than one mapping triplet.